pkg/trie/triedb/node.go

```diff
@@ -168,8 +207,204 @@ func newNodeFromEncoded(nodeHash common.Hash, data []byte, storage NodeStorage)
 			children[i] = child
 		}
 
-		return Branch{partialKey: key, children: children, value: NewFromEncoded(value)}, nil
+		return Branch{partialKey: key, children: children, value: NewValueFromEncoded(encoded.PartialKey, value)}, nil
+	default:
+		panic("unreachable")
+	}
+}
+
+type NodeToEncode interface {
+	isNodeToEncode()
+}
+
+type (
+	NewNodeToEncode struct {
+		partialKey []byte
+		value      []byte
+	}
+	TrieNodeToEncode struct {
+		child NodeHandle
+	}
+)
+
+func (NewNodeToEncode) isNodeToEncode()  {}
+func (TrieNodeToEncode) isNodeToEncode() {}
+
+type ChildReference interface {
+	isChildReference()
+	getNodeData() []byte
+}
+
+type (
+	HashChildReference struct {
+		hash common.Hash
+	}
+	InlineChildReference struct {
+		encodedNode []byte
+	}
+)
+
+func (HashChildReference) isChildReference() {}
+func (h HashChildReference) getNodeData() []byte {
+	return h.hash.ToBytes()
+}
+func (InlineChildReference) isChildReference() {}
+func (i InlineChildReference) getNodeData() []byte {
+	return i.encodedNode
+}
+
+type childFunc = func(node NodeToEncode, partialKey []byte, childIndex *byte) (ChildReference, error)
+
+const firstPrefix = (0x00 << 6)
+const emptyTrieBytes = firstPrefix | (0x00 << 4)
```

Comment:
From what I understood, theses shifts can be replaced by a cohesive
```suggestion
const emptyTrieBytes = []byte{0}
```

Reply:
Actually if we want to replace it should be by:

`const emptyTrieBytes = byte(0)`

Reply:
Replaced :D